classes/lifecycle/step.php

```diff
@@ -49,93 +47,112 @@ public function get_subpluginname() {
-            helper::upload_file($processid, $instanceid, $courseid, $newfile);
-
-            // Delete file.
-            $file->delete();
+        // Set the custom data.
+        $tashhash = md5($processid . '_' . $instanceid . '_' . $course->id);
+        $task->set_custom_data((object) [
+            'processid' => $processid,
+            'stepinstanceid' => $instanceid,
+            'courseid' => $course->id,
+            'workflowid' => $workflowid,
+            'taskhash' => $tashhash,
+        ]);
+
+        // Schedule to run the task after 1 minute.
+        $task->set_next_run_time(time() + 60);
+        \core\task\manager::queue_adhoc_task($task, true);
+
+        // Log the task creation.
+        mtrace("Adhoc course backup task for the course " . $course->id . " has been created");
+
+        // Important, We need to put this to waiting state, so next step has to wait.
+        return step_response::waiting();
+    }
+
+    /**
+     * Only proceed if the course backup task has finished.
+     *
+     * @param int $processid
+     * @param int $instanceid
+     * @param mixed $course
+     * @return step_response
+     */
+    public function process_waiting_course($processid, $instanceid, $course) {
+        if ($this->get_adhoc_task($processid, $instanceid, $course)) {
+            // Keep waiting for the adhoc task to finish.
+            mtrace("Adhoc course backup task for the course " . $course->id . " is still running, waiting.");
+            return step_response::waiting();
+        } else {
+            mtrace("Adhoc course backup task for the course " . $course->id . " has finished, proceeding to next step.");
+            // Task is finished, proceed to next step.
+            return step_response::proceed();
         }
+    }
 
-        // Clean up.
-        $bc->destroy();
-        unset($bc);
+    /**
+     * Rollback the step.
+     * We only delete the adhoc task if it exists
+     *
+     * @param int $processid of the respective process.
+     * @param int $instanceid of the step instance.
+     * @param mixed $course to be rolled back.
+     */
+    public function rollback_course($processid, $instanceid, $course) {
+        global $DB;
+
+        // Find the adhoc task.
+        $task = $this->get_adhoc_task($processid, $instanceid, $course);
+        if ($task) {
+            // Delete the adhoc task.
+            $DB->delete_records('task_adhoc', ['id' => $task->id]);
+        }
+    }
 
-        return step_response::proceed();
+    /**
+     * Find adhoc task
+     *
+     * @param int $processid the process id.
+     * @param int $instanceid step instance id.
+     * @param object $course the course object.
+     *
+     * @return \stdClass|false if the task exists.
+     */
+    private function get_adhoc_task($processid, $instanceid, $course) {
+        global $DB;
+        // Find adhoc task using hash value.
+        $tashhash = md5($processid . '_' . $instanceid . '_' . $course->id);
```

Comment:
This code is being used at least in two spots in this class. Maybe abstract it to a small helper function to avoid code duplication?